lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -0,0 +1,1101 @@
+    ]
+  where
+    tr' = contramap MsgConnectionLost tr
+
+-- | Handlers that are failing if the connection is lost
+failOnConnectionLost
+    :: Tracer IO (NetworkLayerLog c)
+    -> [RetryStatus -> Handler IO Bool]
+failOnConnectionLost tr =
+    [ const $ Handler $ handleIOException tr' False
+    , const $ Handler $ handleMuxError tr' False
+    ]
+  where
+    tr' = contramap MsgConnectionLost tr
+
+-- When the node's connection vanished, we may also want to handle things in a
+-- slightly different way depending on whether we are a waller worker or just
+-- the node's tip thread.
+handleIOException
+    :: Tracer IO (Maybe IOException)
+    -> Bool -- ^ 'True' = retry on 'ResourceVanishedError'
+    -> IOException
+    -> IO Bool
+handleIOException tr onResourceVanished e
+    -- There's a race-condition when starting the wallet and the node at the
+    -- same time: the socket might not be there yet when we try to open it.
+    -- In such case, we simply retry a bit later and hope it's there.
+    | isDoesNotExistError e =
+        pure True
+
+    | isResourceVanishedError e = do
+        traceWith tr $ Just e
+        pure onResourceVanished
+
+    | otherwise =
+        pure False
+  where
+    isResourceVanishedError = isInfixOf "resource vanished" . show
+
+handleMuxError
+    :: Tracer IO (Maybe IOException)
+    -> Bool -- ^ 'True' = retry on 'ResourceVanishedError'
+    -> MuxError
+    -> IO Bool
+handleMuxError tr onResourceVanished = pure . errorType >=> \case
+    MuxUnknownMiniProtocol -> pure False
+    MuxDecodeError -> pure False
+    MuxIngressQueueOverRun -> pure False
+    MuxInitiatorOnly -> pure False
+    MuxSDUReadTimeout -> pure False
+    MuxSDUWriteTimeout -> pure False
+    MuxIOException e ->
+        handleIOException tr onResourceVanished e
+    MuxBearerClosed -> do
+        traceWith tr Nothing
+        pure onResourceVanished
+
+-- | Client for the 'Chain Sync' mini-protocol.
+--
+-- A corresponding 'Channel' can be obtained using a `MuxInitiatorApplication`
+-- constructor. Once started, the client simply runs ad-infinitum but one may
+-- interact with it via a 'TQueue' of commands / messages used to move inside
+-- the state-machine.
+--
+-- In a typical usage, 'chainSyncWithBlocks' would be executed in a forked
+-- thread and given a 'TQueue' over which the parent thread as control.
+--
+-- >>> forkIO $ void $ chainSyncWithBlocks tr queue channel
+-- ()
+-- >>> writeTQueue queue ...
+--
+--                                    Agency
+--     -------------------------------------------------------------------------
+--     Client has agency*                | Idle
+--     Server has agency*                | Intersect, Next
+--
+--     * A peer has agency if it is expected to send the next message.
+--
+--      *-----------*
+--      | Intersect |◀══════════════════════════════╗
+--      *-----------*         FindIntersect         ║
+--            │                            :750
```

Comment:
:thinking:

Reply:
👍Fixed and changed the commits to be more
1. Copy byron implementation
2. Tweak

for nicer diffs.